share/wake/lib/core/json.wake

```diff
@@ -169,3 +169,143 @@ export def x ==/ y = match x y
+    JSONNormalize Pass Pass Pass Pass Pass Pass
+
+# Target the minimum JSON language described by the specifications.
+export def normalizeJSONCompat: JSONNormalize =
+    def nfkcKeys =
+        map (editPairFirst unicodeIdentifier)
+    normalizeJSONIdentity
+    | setJSONNormalizeString (Pass _.unicodeCanonical)
+    | setJSONNormalizeDouble filterNonDigitJDouble
+    | setJSONNormalizeObject (deduplicateJObjectKeys _.nfkcKeys)
+
+# Fail on any `Double` values which can't be represented numerically.
+# Published JSON specifications describe their double values as allowing decimal
+# or exponential forms, but don't implement the full IEEE standard.
+def filterNonDigitJDouble (n: Double): Result Double Error = match n.dclass
+    DoubleInfinite =
+        failWithError "For compatibility, JSON doubles should not be infinite."
+    DoubleNaN =
+        failWithError "For compatibility, JSON doubles should not be NaN values."
+    _ =
+        Pass n
+
+# Attempt to simplify any `JObject`s which contain multiple instances of a key.
+# While the published specifications explicitly allow such duplicate keys, none
+# *require* implementations to accept them nor describes the behaviour by which
+# they should be handled; therefore, implementations differ in how they handle
+# duplicate keys -- if they do so at all -- and a generator aiming for
+# compatibility shouldn't output objects with duplicate keys.
+#
+# In particular, this function will concatenate the members of two `JArray`
+# values, treat `JNull` as a wildcard placeholder, and otherwise return the
+# original value if both instances are equal. It fails if a key is associated
+# with multiple values of different types or with different values of a type
+# which cannot be combined.
+def deduplicateJObjectKeys (dict: List (Pair String JValue)): Result (List (Pair String JValue)) Error =
+    def cmpKeysOnly (Pair k1 _) (Pair k2 _) =
+        k1 <* k2
+    def simplifyKeyGroups = match _
+        Nil =
+            # `simplifyKeyGroups` is only ever mapped over the output of
+            # `groupBy`, which is always a list of *non-empty* lists.
+            failWithError "groupBy invariant reached in deduplicateJObjectKeys"
+        (Pair key value), ps =
+            (value, map getPairSecond ps)
+            | rfoldr (simplifyValues key) JNull
+            | rmap (Pair key)
+    def simplifyValues key: JValue => JValue => Result JValue Error = match _ _
+        val JNull =
+            Pass val
+        JNull acc =
+            Pass acc
+        (JString val) (JString acc) if val ==* acc =
+            Pass (JString acc)
+        (JInteger val) (JInteger acc) if val == acc =
+            Pass (JInteger acc)
+        (JDouble val) (JDouble acc) if val ==. acc =
+            Pass (JDouble acc)
+        (JBoolean val) (JBoolean acc) if enor val acc =
+            Pass (JBoolean acc)
+        (JArray val) (JArray acc) =
+            Pass (JArray (val ++ acc))
```

Comment:
> I think you're overthinking this one. Your rule is too magical. It can switch behaviours for a user-inexplicable reason. Pick one behaviour, only. I'd just go for `JArray (val ++ acc)` unconditionally, since that's the more helpful of the two options.

I went with that because then `deduplicateJObjectKeys (x ++ x) == deduplicateJObjectKeys x` (i.e. merging an object into itself wouldn't change anything), but it is admittedly rather magical. The only other solution I have would be to filter out duplicates in the second array (union of the two), but that's probably too prone to errors such as deliberately-duplicated values. Ultimately, I've just wound up going with your suggestion for an unconditional append.

Reply:
I'm not sure we want this property (`x ++ x == x`).

Reply:
It's definitely not something we'd want to pursue, but just due to how everything else works, everything but arrays coincidentally happens to have that property.  It's basically weighing which rule we want the exception on (i.e. "arrays append *except when*" vs "it's idempotent *except for*"), and I do think you're right that the latter's the better way to go -- especially since if we never point it out in the docs, there's probably not going to be as many people noticing that one on their own.

Reply:
Yeah. Also for consistency, I think since we "append" objects, we should "append" arrays.